src/Storages/MergeTree/MergeTreeIndexAnnoy.cpp

```diff
@@ -0,0 +1,290 @@
+void AnnoyIndexSerialize<Dist>::deserialize(ReadBuffer& istr)
+{
+    readIntBinary(Base::_s, istr);
+    readIntBinary(Base::_n_items, istr);
+    readIntBinary(Base::_n_nodes, istr);
+    readIntBinary(Base::_nodes_size, istr);
+    readIntBinary(Base::_K, istr);
+    readIntBinary(Base::_seed, istr);
+    readVectorBinary(Base::_roots, istr);
+    Base::_nodes = realloc(Base::_nodes, Base::_s * Base::_n_nodes);
+    istr.read(reinterpret_cast<char*>(Base::_nodes), Base::_s * Base::_n_nodes);
+
+    Base::_fd = 0;
+    // set flags
+    Base::_loaded = false;
+    Base::_verbose = false;
+    Base::_on_disk = false;
+    Base::_built = true;
+}
+
+template<typename Dist>
+float AnnoyIndexSerialize<Dist>::getSpaceDim() const
+{
+    return Base::get_f();
+}
+
+}
+
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int INCORRECT_QUERY;
+}
+
+MergeTreeIndexGranuleAnnoy::MergeTreeIndexGranuleAnnoy(const String & index_name_, const Block & index_sample_block_)
+    : index_name(index_name_)
+    , index_sample_block(index_sample_block_)
+    , index_base(nullptr)
+{}
+
+MergeTreeIndexGranuleAnnoy::MergeTreeIndexGranuleAnnoy(
+    const String & index_name_,
+    const Block & index_sample_block_,
+    AnnoyIndexPtr index_base_)
+    : index_name(index_name_)
+    , index_sample_block(index_sample_block_)
+    , index_base(std::move(index_base_))
+{}
+
+bool MergeTreeIndexGranuleAnnoy::empty() const
+{
+    return !static_cast<bool>(index_base);
+}
+
+void MergeTreeIndexGranuleAnnoy::serializeBinary(WriteBuffer & ostr) const
+{
+    writeIntBinary(index_base->getSpaceDim(), ostr); // write dimension
+    index_base->serialize(ostr);
+}
+
+void MergeTreeIndexGranuleAnnoy::deserializeBinary(ReadBuffer & istr, MergeTreeIndexVersion /*version*/)
+{
+    int dimension;
+    readIntBinary(dimension, istr);
+    index_base = std::make_shared<AnnoyIndex>(dimension);
+    index_base->deserialize(istr);
+}
+
+
+MergeTreeIndexAggregatorAnnoy::MergeTreeIndexAggregatorAnnoy(const String & index_name_,
+                                                                const Block & index_sample_block_,
+                                                                int index_param_)
+    : index_name(index_name_)
+    , index_sample_block(index_sample_block_)
+    , index_param(index_param_)
+{}
+
+bool MergeTreeIndexAggregatorAnnoy::empty() const
+{
+    return !index_base || index_base->get_n_items() == 0;
+}
+
+MergeTreeIndexGranulePtr MergeTreeIndexAggregatorAnnoy::getGranuleAndReset()
+{
+    index_base->build(index_param);
+    auto granule = std::make_shared<MergeTreeIndexGranuleAnnoy>(index_name, index_sample_block, index_base);
+    index_base = nullptr;
+    return granule;
+}
+
+void MergeTreeIndexAggregatorAnnoy::update(const Block & block, size_t * pos, size_t limit)
+{
+    if (*pos >= block.rows())
+        throw Exception(
+                "The provided position is not less than the number of block rows. Position: "
```

Comment:
One can use libfmt-style placeholder syntax. Just as an example from another file:

```
throw Exception(ErrorCodes::NOT_IMPLEMENTED, "updateImpl() is not implemented in {}", getStorageType());
```

Reply:
In many indexes we have same lines. Example https://github.com/ClickHouse/ClickHouse/blob/2a1ede0f5ab2bdefaca3d2c05ab659ecd66ec75f/src/Storages/MergeTree/MergeTreeIndexMinMax.cpp#L124-L127

I also prefer libfmt-style placeholder syntax, I'll change it soon. Probably all indexes should have